rclcpp_action/src/client.cpp

```diff
@@ -12,30 +12,350 @@
+    wait_set, pimpl_->client_handle.get(), nullptr, nullptr);
+  return RCL_RET_OK == ret;
+}
+
+bool
+ClientBase::is_ready(rcl_wait_set_t * wait_set)
+{
+  rcl_ret_t ret = rcl_action_client_wait_set_get_entities_ready(
+    wait_set, pimpl_->client_handle.get(),
+    &pimpl_->is_feedback_ready,
+    &pimpl_->is_status_ready,
+    &pimpl_->is_goal_response_ready,
+    &pimpl_->is_cancel_response_ready,
+    &pimpl_->is_result_response_ready);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(
+      ret, "failed to check for any ready entities");
+  }
+  return
+    pimpl_->is_feedback_ready ||
+    pimpl_->is_status_ready ||
+    pimpl_->is_goal_response_ready ||
+    pimpl_->is_cancel_response_ready ||
+    pimpl_->is_result_response_ready;
+}
+
+void
+ClientBase::handle_goal_response(
+  const rmw_request_id_t & response_header,
+  std::shared_ptr<void> response)
+{
+  std::lock_guard<std::mutex> guard(pimpl_->goal_requests_mutex);
+  const int64_t & sequence_number = response_header.sequence_number;
+  if (pimpl_->pending_goal_responses.count(sequence_number) == 0) {
+    RCLCPP_ERROR(pimpl_->logger, "unknown goal response, ignoring...");
+    return;
+  }
+  pimpl_->pending_goal_responses[sequence_number](response);
+  pimpl_->pending_goal_responses.erase(sequence_number);
+}
+
+void
+ClientBase::send_goal_request(std::shared_ptr<void> request, ResponseCallback callback)
+{
+  std::unique_lock<std::mutex> guard(pimpl_->goal_requests_mutex);
+  int64_t sequence_number;
+  rcl_ret_t ret = rcl_action_send_goal_request(
+    pimpl_->client_handle.get(), request.get(), &sequence_number);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret, "failed to send goal request");
+  }
+  assert(pimpl_->pending_goal_responses.count(sequence_number) == 0);
+  pimpl_->pending_goal_responses[sequence_number] = callback;
+}
+
+void
+ClientBase::handle_result_response(
+  const rmw_request_id_t & response_header,
+  std::shared_ptr<void> response)
+{
+  std::lock_guard<std::mutex> guard(pimpl_->result_requests_mutex);
+  const int64_t & sequence_number = response_header.sequence_number;
+  if (pimpl_->pending_result_responses.count(sequence_number) == 0) {
+    RCLCPP_ERROR(pimpl_->logger, "unknown result response, ignoring...");
+    return;
+  }
+  pimpl_->pending_result_responses[sequence_number](response);
+  pimpl_->pending_result_responses.erase(sequence_number);
+}
+
+void
+ClientBase::send_result_request(std::shared_ptr<void> request, ResponseCallback callback)
+{
+  std::lock_guard<std::mutex> guard(pimpl_->result_requests_mutex);
+  int64_t sequence_number;
+  rcl_ret_t ret = rcl_action_send_result_request(
+    pimpl_->client_handle.get(), request.get(), &sequence_number);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret, "failed to send result request");
+  }
+  assert(pimpl_->pending_result_responses.count(sequence_number) == 0);
```

Comment:
Why can't we have multiple pending result responses?

Reply:
I think this is asserting that there is only one pending response with the same sequence number

Reply:
Ah yes, misread that.